keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,364 @@
+
+### Inline CSI volume operation stages
+When a CSI driver is used in an inline context, it works slightly differently then when originated from PV/PVC.  As mentioned earlier, the inline volumes will participate in some, but not all volume operation stages with some limitations discussed here.
+
+#### Provision/deletion
+Volume provision and deletion will work differently for inline volumes. For persistent inline volumes, provision/deletion is handled as follows:
+* Persistent inline volumes do not participate in the provision and deletion phases.
+* Persistent inline volumes expect provision and deletion to be handled outside of the driver.
+
+For ephemeral CSI drivers:
+* Ephemeral drivers will not receive provision/deletion API calls (since these stages are driven by a PV/PVC/StorageClass).
+* Ephemeral CSI drivers will have to delay or combine any provisioning/deprovisioning operation during different phase.
+
+#### Attach/detach
+CSI uses API object `storage.VolumeAttachment` to track and manage attach/detach storage operation. Currently that object contains reference to an associated PV when the driver is not used inline.  However,  it must be extended to support information from `CSIVolumeSource` (see above) when an inline volume is being attached.
+
+```go
+// VolumeAttachmentSpec is the specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+    // <snip>
+
+	// Source represents the volume that should be attached.
+	Source VolumeAttachmentSource
+}
+
+// VolumeAttachmentSource represents a volume that should be attached, either
+// PersistentVolume or a volume in-lined in a Pod.
+// Exactly one member can be set.
+type VolumeAttachmentSource struct {
+	// Name of the persistent volume to attach.
+	// +optional
+	PersistentVolumeName *string
+
+	// InlineVolumeSource represents the source location of a in-line volume in a pod to attach.
+	// +optional
+    InlineVolumeSource *InlineVolumeSource
+}
+
+// InlineVolumeSource represents the source location of a in-line volume in a pod.
+type InlineVolumeSource struct {
+	// VolumeSource is copied from the pod. It ensures that attacher has enough
+	// information to detach a volume when the pod is deleted before detaching.
+	// Only CSIVolumeSource can be set.
+	// Required.
+	CSIVolumeSource v1.VolumeSource
+
+	// Namespace of the pod with in-line volume. It is used to resolve
+	// references to Secrets in VolumeSource.
+	// Required.
+	Namespace string
+}
+```
+The following steps outline how the API is used to track attachment and detachment of volumes:
+* The external CSI A/D controller **copies whole `VolumeSource`**  from `Pod` into `VolumeAttachment`. This allows external CSI attacher to detach volumes for deleted pods without keeping any internal database of attached VolumeSources.
```

Comment:
@liggitt this is a very good point.  

>if so, what occurs when two pods (on the same node, I guess) define the same volumeHandle inline with different volume source info?

 If a second pod defines a volumeHandle with the same name as a previously defined handle in a different pod and both happen to land on the same machine:
- The second attachment request will be ignored since it's already there
- There will be a race where a detachment may come before one of the pods is ready for detachment
- In that case, new code will be needed to enumerate attachments and ensure that attachment are marked ready for detachment only when the last pod is done.
- I will update the KEP to point that out.

> if the pods are in different namespaces, can deletion of one namespace break attach/detach for the other pod if there are inaccessible secret references in the first pod's namespace?

This case is documented in the KEP already.  If access to a namepsace, needed for volume operations, is broken then that volume operation will fail and communicated to the user.

Reply:
> if so, what occurs when two pods (on the same node, I guess) define the same volumeHandle inline with different volume source info?

I will also investigate the possibility of catching that condition using validation.  Not sure if that is possible right now.

Reply:
> * In that case, new code will be needed to enumerate attachments and ensure that attachment are marked ready for detachment only when the last pod is done.

A/D controller already does refcnting if multiple pods share the same volume.  I'm not sure if we need to do anything special here.  The main problem that we need to figure out is if two pods have conflicting information that impacts attach, like the namespace.

Reply:
> There will be a race where a detachment may come before one of the pods is ready for detachment

Does that match behavior of existing inline volumes (ebs or gcepd volumes, etc)? Or were you saying that's an issue that needs to be resolved as part of this?

> If access to a namepsace, needed for volume operations, is broken then that volume operation will fail and communicated to the user.

For a single pod, I guess that makes sense (user1 creates a pod with an inline persistent volume foo requiring secret `bar` in their namespace to detach, then deletes secret `bar` and sees errors when attempting to detach)

For two pods from different namespaces, that's:
a) potentially very difficult to understand (user 2 creates a pod with an inline persistent volume and start seeing errors about secrets they didn't reference in a namespace that isn't theirs. also, would the detach errors even get reported to user 2's namespace?)
b) impossible for user 2 to recover from (there's no way to switch the volumesource/namespace to a still-existing pod/namespace)
c) concerning, security-wise, if user 2 can go fishing for a volumeHandle that is already attached on their node and get it mounted into their pod without needing to provide the required attach/mount secrets

What is the behavior of existing inline volumes around these aspects? It would help to know if these are new issues, or recognition of existing issues

Reply:
@liggitt issues in a) - c) are most likely present in existing inline volumes.  But to be clear ...

>a) ... user 2 creates a pod with an inline persistent volume and start seeing errors about secrets they didn't reference in a namespace that isn't theirs.

... I will restate the above:
User 2's pod would make reference to a secret that is part of a namespace that user 2 does not own.  Then that namespace disappear.  During detach, for instance, the external AD attempts to detach but, the missing namespace causes error.

I would point out that the `volumesource` uses `LocalObjectReference` which causes the code to assume that all referenced resources are part of the the pod's namespace.


Lastly...
> c) concerning, security-wise, if user 2 can go fishing for a volumeHandle that is already attached on their node and get it mounted into their pod without needing to provide the required attach/mount secrets

Well if user 2 happens to guess or obtain the right volumeHandle value and also have access to the namespace where secrets (to access the driver) are defined,  then yes, all bets are off.

Reply:
@liggitt 
> A correctly specified user1 pod in ns1 and user2 pod in ns2 exist, both scheduled to the same node. Each pod has a complete volumesource, and refers to valid secrets in their own namespace.

I am guessing in this scenario user1 pod and user2 pod use the same volumeHandle? That would be the only time the volumeattachment would have an ID shared by both pods during A/D.  Otherwise, it's a different volumeattachment object, no issue.

> User 2 now has a pod with a valid volume source and valid secrets, serviced by a volumeattachment object pointing at a defunct namespace/secrets. How does the attach/detach controller know to tell user 2 about the attach/detach problems? How does user 2 recover from this scenario?

So yes, if user1 wins the attachment race, its info will be inside volumeattachment. If user1 deletes its namespace/secrets, then during detachment, if we assumes that user2 pod is going to use that same attachment object, it will fail.  The reporting mechanism right now is K8s events.

This discussion reveals that it may be a good idea to limit 1 attachment object per inline volume pod spec (not sure how yet).  

I would imagine these issues are not all resolved in existing in-tree inline drivers.  Using a more comprehensive validation that takes into account pods with persistent inline csi and specified volhandle value maybe something to explore to avoid these scenarios.

Reply:
I will update the doc to point out these issues and how they may be handled.